File: src/advent_of_code/2022/day18.clj
(ns advent-of-code.2022.day18
  "Day 18: Boiling Boulders https://adventofcode.com/2022/day/18"
  (:require
   [clojure.string :as str]
   [clojure.set]))


(def input-filename "inputs/2022/day18.txt")

(def set-all-cubes
  (->> (slurp input-filename)
       str/split-lines
       (map (fn [line] (->> (str/split line #",")
                            (mapv #(Integer/parseInt %)))))
       (into #{})))

(defn add-coords [coords-a coords-b]
  (map + coords-a coords-b))

;; surrounding only means the ones touching each face
;;     NOT diagonally
(defn surrounding-cubes-coords [cube-coord]
  (let [offsets  [[1 0 0] [-1 0 0]
                  [0 1 0] [0 -1 0]
                  [0 0 1] [0 0 -1]]]
    (map (partial add-coords cube-coord) offsets)))

(defn exposed-sides-count [set-all-cubes cube-coord]
  (let [covered-faces-count (->> cube-coord
                                 surrounding-cubes-coords
                                 (filter set-all-cubes)
                                 count)]
    (- 6 covered-faces-count)))

(assert (= 5  (let [cube-coord [1 1 1]
                    set-all-cubes #{[1 1 1] [2 1 1]}]
                (exposed-sides-count set-all-cubes cube-coord))))

(->> set-all-cubes
     (map (partial exposed-sides-count set-all-cubes))
     (apply +))
;; answer 1: 4504

;; PART 2

;; imagine a cuboid which completely surrounds the entire lava droplet (with minimum 1 "padding" in all directions")
;; because of the padding, we can consider one corner point as the starting point and then 
;;   breadth first traversal into all possible points to get the `set-of-all-points-which-can-be-reached-by-water-or-steam`
;; then we can easily get `coords-surrounding-lava-which-cant-be-reached` and use that to figure out external surface area
(let [[x-max y-max z-max]
      (->> set-all-cubes
           (apply map max)
           (map inc))
      [x-min y-min z-min]
      (->> set-all-cubes
           (apply map min)
           (map dec))
      within-bounds-fn (fn [[coord-x coord-y coord-z]]
                         (and (<= x-min coord-x x-max)
                              (<= y-min coord-y y-max)
                              (<= z-min coord-z z-max)))
      starting-point [x-min y-min z-min]
      ;; breadth first traversal 
      set-of-all-points-which-can-be-reached-by-water-or-steam
      (loop [visited-set #{}
             frontier    #{starting-point}]
        #_(println "loop: (count frontier):" (count frontier))
        (if (empty? frontier)
          visited-set
          (let [new-frontier*   (->> frontier
                                     (mapcat surrounding-cubes-coords)
                                     (filter within-bounds-fn)
                                     (into #{}))
                new-frontier    (clojure.set/difference new-frontier* visited-set set-all-cubes)
                new-visited-set (clojure.set/union visited-set frontier)]
            (recur new-visited-set new-frontier))))
      lava-coord->externally-exposed-sides-count
      (fn [lava-coord]
        (let [surrounding-coords-set (into #{} (surrounding-cubes-coords lava-coord))
              surrounding-coords-exposed-externally
              ;; don't need to account for coords in set-all-cubes here because they are not contained in `set-of-all-points-which-can-be-reached-by-water-or-steam` anyways
              (clojure.set/intersection surrounding-coords-set
                                        set-of-all-points-which-can-be-reached-by-water-or-steam)]
          (count surrounding-coords-exposed-externally)))]
  (->> set-all-cubes
       (map lava-coord->externally-exposed-sides-count)
       (apply +)))
;; answer 2: 2556

File: python/2022/day18.py
"""Day 18: Boiling Boulders - https://adventofcode.com/2022/day/18"""

from collections import deque

INPUT_FILE = "inputs/2022/day18.txt"


def parse_input(filename: str) -> set[tuple[int, int, int]]:
    """Parse cube coordinates."""
    cubes = set()
    with open(filename) as f:
        for line in f:
            x, y, z = map(int, line.strip().split(","))
            cubes.add((x, y, z))
    return cubes


def get_neighbors(coord: tuple[int, int, int]) -> list[tuple[int, int, int]]:
    """Get 6 face-adjacent neighbors."""
    x, y, z = coord
    return [
        (x + 1, y, z), (x - 1, y, z),
        (x, y + 1, z), (x, y - 1, z),
        (x, y, z + 1), (x, y, z - 1)
    ]


def count_exposed_faces(cubes: set[tuple[int, int, int]]) -> int:
    """Count faces not touching another cube."""
    total = 0
    for cube in cubes:
        for neighbor in get_neighbors(cube):
            if neighbor not in cubes:
                total += 1
    return total


def count_external_faces(cubes: set[tuple[int, int, int]]) -> int:
    """Count faces exposed to exterior (not trapped air pockets)."""
    # Find bounding box with padding
    if not cubes:
        return 0

    min_x = min(c[0] for c in cubes) - 1
    max_x = max(c[0] for c in cubes) + 1
    min_y = min(c[1] for c in cubes) - 1
    max_y = max(c[1] for c in cubes) + 1
    min_z = min(c[2] for c in cubes) - 1
    max_z = max(c[2] for c in cubes) + 1

    def in_bounds(coord):
        x, y, z = coord
        return (min_x <= x <= max_x and
                min_y <= y <= max_y and
                min_z <= z <= max_z)

    # BFS from corner to find all reachable external points
    start = (min_x, min_y, min_z)
    visited = {start}
    queue = deque([start])

    while queue:
        current = queue.popleft()
        for neighbor in get_neighbors(current):
            if neighbor not in visited and neighbor not in cubes and in_bounds(neighbor):
                visited.add(neighbor)
                queue.append(neighbor)

    # Count cube faces touching external air
    external_faces = 0
    for cube in cubes:
        for neighbor in get_neighbors(cube):
            if neighbor in visited:
                external_faces += 1

    return external_faces


def part1(cubes: set[tuple[int, int, int]]) -> int:
    """Total surface area."""
    return count_exposed_faces(cubes)


def part2(cubes: set[tuple[int, int, int]]) -> int:
    """External surface area only."""
    return count_external_faces(cubes)


def main():
    cubes = parse_input(INPUT_FILE)

    answer1 = part1(cubes)
    answer2 = part2(cubes)

    print(f"Part 1: {answer1}")  # 4504
    print(f"Part 2: {answer2}")  # 2556

    return answer1, answer2


if __name__ == "__main__":
    main()
